pipeline/transform.py

```diff
@@ -0,0 +1,299 @@
+            tags = soup.find_all('a', class_='tag')
+
+            if tags:
+                genres = [tag.text for tag in tags if tag.text.lower()
+                          not in locations]
+                genres = list(set(genres))
+                return genres
+
+            logging.info("No genres listed for %s", full_url)
+            return []
+
+        logging.warning("Failed to connect to %s, Status Code: %s",
+                        full_url, response.status_code)
+        return []
+
+    except requests.exceptions.RequestException as e:
+        logging.error("Request failed for %s: %s", artist_url, e)
+
+    except Exception as e:
+        logging.error("An unexpected error occurred: %s", e)
+
+    return []
+
+
+def get_release_date_from_url(artist_url: str) -> str:
+    """
+    Retrieves the release date of a song or album provided in the URL.
+    Returns an empty string if it cannot be found or an error occurs.
+    """
+    try:
+        full_url = convert_to_full_url(artist_url)
+        response = requests.get(full_url, timeout=1000)
+
+        if response.status_code == 200:
+            logging.info("Getting release date for %s", full_url)
+            soup = BeautifulSoup(response.text, 'html.parser')
+            meta_tag = soup.find('meta', attrs={'name': 'description'})
+
+            if meta_tag:
+                description_content = meta_tag.get('content')
+                if description_content:
+                    date_pattern = r'released\s(\d{1,2}\s[a-zA-Z]+\s\d{4})'
+                    match = re.search(date_pattern, description_content)
+                    if match:
+                        release_date = match.group(1)
+                        return convert_date_format(release_date)
+
+            logging.info("No valid release date found for %s", full_url)
+
+        else:
+            logging.warning("Failed to connect to %s, Status Code: %s",
+                            full_url, response.status_code)
+
+    except requests.exceptions.RequestException as e:
+        logging.error("Request failed for %s: %s", artist_url, e)
+
+    except Exception as e:
+        logging.error("An unexpected error occurred: %s", e)
+
+    return ""
+
+
+def extend_sales_from_df(sales_df: pd.DataFrame) -> None:
+    """
+    Coverts the "sale_date" column into datetime format 
+    and extends the given dataframe with month and year
+    of sales.
+    Does nothing if a "sale_date" column doesn't exist.
+    """
+
+    if "sale_date" in sales_df.columns:
+        sales_df["sale_date"] = pd.to_datetime(
+            sales_df["sale_date"], errors="coerce")
+
+        if sales_df["sale_date"].isna().any():
+            logging.warning(
+                "Some 'sale_date' entries could not be converted to datetime.")
+
+    sales_df["sale_year"] = sales_df["sale_date"].dt.year
+    sales_df["sale_month"] = sales_df["sale_date"].dt.month
+
+
+def replace_blank_album_titles(sales_df: pd.DataFrame) -> None:
```

Comment:
Don't we want to do this with tracks? My understanding was that tracks commonly had their title in the 'item_description' field.

Reply:
It was albums that had their name in the `item_description` column, but was lacking it in the `album_title` column.

Tracks always have their name in `item_description` but things like singles don't necessarily have an associated album, so it would be right to have some "None" values for them in the table.